File: python3/asteroid_collision.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for e in asteroids:
            if not stack or stack[-1] < 0 or e > 0:
                stack.append(e)

            else:
                flag = True
                while stack and stack[-1] > 0 and stack[-1] <= -e:
                    tmp = stack.pop()
                    if tmp == -e:
                        flag = False
                        break
                if (not stack or stack[-1] < 0) and flag:
                    stack.append(e)
        return stack

File: golang/asteroid_collision.go
package main

/*
Space	: O(n)
Time	: O(n)
*/

func asteroidCollision(asteroids []int) []int {
	stack := []int{}

	for _, ast := range asteroids {
		if len(stack) == 0 {
			stack = append(stack, ast)
			continue
		}

		push := true
		for len(stack) > 0 && stack[len(stack)-1] > 0 && ast < 0 {
			if stack[len(stack)-1] > -ast {
				push = false
				break
			} else if stack[len(stack)-1] < -ast {
				stack = stack[:len(stack)-1]
			} else {
				stack = stack[:len(stack)-1]
				push = false
				break
			}
		}

		if push {
			stack = append(stack, ast)
		}
	}

	return stack
}
